packages/codemirror-blocks/src/ui/PrimitiveList.tsx

```diff
@@ -93,69 +93,139 @@ export const PrimitiveGroup = (props: PrimitiveGroupProps) => {
       {expanded ? (
         <PrimitiveList
           primitives={[...group.flatPrimitivesIter()]}
-          onFocus={onFocus}
-          onKeyDown={onKeyDown}
           selected={selected}
+          selectedPrimitive={selectedPrimitive}
+          toolbarRef={toolbarRef}
+          setSelectedPrimitive={setSelectedPrimitive}
         />
       ) : null}
     </li>
   );
 };
 
 type PrimitiveListProps = {
-  onFocus: BasePrimitiveProps["onFocus"];
-  onKeyDown: BasePrimitiveProps["onKeyDown"];
+  setSelectedPrimitive: (primitive: LanguagePrimitive) => void;
+  selectedPrimitive: LanguagePrimitive | undefined;
+  toolbarRef: React.RefObject<HTMLInputElement>;
   selected?: string;
   primitives?: LanguagePrimitive[];
   searchString?: string;
 };
-export const PrimitiveList = (props: PrimitiveListProps) => {
-  const { primitives = [], selected, onFocus, onKeyDown, searchString } = props;
-  const renderGroup = (g: PrimitiveGroupModel) => (
-    <PrimitiveGroup
-      key={g.name}
-      group={g}
-      onFocus={onFocus}
-      onKeyDown={onKeyDown}
-      selected={selected}
-    />
-  );
-  const renderPrimitive = (p: LanguagePrimitive) => (
-    <Primitive
-      key={p.name}
-      primitive={p}
-      onFocus={onFocus}
-      onKeyDown={onKeyDown}
-      className={selected == p.name ? "selected" : ""}
-    />
-  );
+export const PrimitiveList = React.forwardRef(
+  (props: PrimitiveListProps, ref: React.MutableRefObject<HTMLElement>) => {
+    const {
+      primitives = [],
+      selected,
+      setSelectedPrimitive,
+      selectedPrimitive,
+      toolbarRef,
+      searchString,
+    } = props;
 
-  const text = searchString
-    ? (primitives.length == 0 ? "No" : primitives.length) + " blocks found"
-    : "blocks";
+    // @pcardune - typing this array has been a challenge
+    // JSX.Element doesn't work, even though that's the
+    // recommendation I'm seeing online
+    const primitiveRefs = useRef<any[]>([]);
```

Comment:
This is what we talked about _not_ doing. The refs shouldn't be holding react elements, they should be holding _dom elements_. And instead of having a ref holding an array, have an array holding refs.